File: packages/core/src/contacts.ts
/**
 * contacts.ts — pull contact routes out of a company's own website HTML.
 * Pure (takes an HTML string; no I/O).
 *
 *   const c = extractContactInfo(html, "https://acme.co.uk/", "acme.co.uk");
 *   // -> { contact_phone, contact_phones, contact_email, contact_emails_found,
 *   //      contact_page, about_page, team_page }
 *   lead.best_contact_route = bestContactRoute({ ...lead, ...c });
 *
 * Phone numbers are parsed via libphonenumber-js (default region GB) and stored
 * in two parallel arrays: display ("020 7946 0123") and normalised E.164
 * ("+442079460123"). De-dup is by E.164, so "020 7946 0123" and "+44 20 7946 0123"
 * collapse to one entry.
 *
 * Emails are handled in two passes:
 *   1. Cloudflare email-obfuscation deobfuscator runs over the raw HTML first
 *      (data-cfemail="..." → real address).
 *   2. "[at]" / "[dot]" / "(at)" patterns are normalised before regex extraction.
 */
import * as cheerio from "cheerio";
// /max imports the full metadata bundle (~140 KB) so getType() returns
// "MOBILE"/"FIXED_LINE"/"TOLL_FREE" etc. The default "min" build only validates,
// it doesn't classify — and "is this a switchboard or a personal mobile?" is
// useful enough to be worth the bytes (we only run this server-side anyway).
import { parsePhoneNumberFromString, type PhoneNumber } from "libphonenumber-js/max";

const FINANCE_PREFIXES = ["accounts", "finance", "ap", "ar", "payments", "treasury", "purchasing", "procurement"];
const GENERIC_PREFIXES = ["info", "enquiries", "enquiry", "hello", "contact", "office", "admin", "sales", "mail", "team"];
const JUNK_EMAIL_HOSTS = ["sentry.io", "wixpress.com", "example.com", "domain.com", "email.com", "yourdomain.com", "sentry-next.wixpress.com"];

const PAGE_KEYWORDS: Record<"contact_page" | "about_page" | "team_page", string[]> = {
  contact_page: ["contact", "contact-us", "contactus", "get-in-touch"],
  about_page: ["about", "about-us", "aboutus", "who-we-are", "our-story", "our-company", "the-company"],
  team_page: ["team", "our-team", "meet-the-team", "people", "our-people", "management", "leadership", "directors", "staff", "key-people"],
};

// Wide-net phone regex — anything that looks like 9+ digits separated by typical
// punctuation. libphonenumber-js does the real validation downstream.
const PHONE_RE = /(?<![\w/])(?:\+?\d[\d\s().\-]{8,18}\d)(?![\w/])/g;
const EMAIL_RE = /[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/g;

/**
 * Reverse Cloudflare's email obfuscation. `<a data-cfemail="xxxx...">` encodes
 * the address by XOR'ing each byte with the first byte (the "key"). Idempotent —
 * inputs without data-cfemail come back unchanged.
 */
export function deobfuscateCloudflareEmails(html: string): string {
  return html.replace(/data-cfemail=["']([a-f0-9]+)["']/gi, (full, hex: string) => {
    if (!hex || hex.length < 4 || hex.length % 2) return full;
    const key = parseInt(hex.slice(0, 2), 16);
    let out = "";
    for (let i = 2; i < hex.length; i += 2) {
      const code = parseInt(hex.slice(i, i + 2), 16) ^ key;
      if (Number.isNaN(code)) return full;
      out += String.fromCharCode(code);
    }
    return `data-cfemail="${hex}">${out}<`;
  });
}

/** Normalise common "[at]" / "[dot]" / "(at)" obfuscations to "@" / "." so EMAIL_RE catches them. */
export function deobfuscateAtDot(text: string): string {
  return text
    .replace(/\s*[[(\{]\s*at\s*[\])\}]\s*/gi, "@")
    .replace(/\s*[[(\{]\s*dot\s*[\])\}]\s*/gi, ".")
    .replace(/\s+at\s+(?=[a-z0-9][a-z0-9.-]*\s+dot\s+[a-z]{2,})/gi, "@")    // " name at acme dot com"
    .replace(/(?<=@[a-z0-9-]+(?:\.[a-z0-9-]+)*)\s+dot\s+(?=[a-z]{2,})/gi, ".");
}

/**
 * Parse a raw phone string with libphonenumber-js (default region GB).
 * Returns { display, e164, type } or null if it doesn't look like a real number.
 *   - display: normalised national format ("020 7946 0123")
 *   - e164: "+442079460123" (used for dedup)
 *   - type: "fixed_line" | "mobile" | "toll_free" | "premium_rate" | … or "" if unknown
 */
export function parsePhone(raw: string, defaultRegion: "GB" | "US" | "IE" | "FR" | "DE" = "GB"):
  | { display: string; e164: string; type: string }
  | null
{
  if (!raw) return null;
  let parsed: PhoneNumber | undefined;
  try { parsed = parsePhoneNumberFromString(raw, defaultRegion); } catch { return null; }
  if (!parsed || !parsed.isPossible()) return null;
  const type = (parsed.getType?.() ?? "").toString().toLowerCase();
  return { display: parsed.formatNational(), e164: parsed.number, type };
}

function registrable(host: string): string {
  return (host || "").toLowerCase().split(":")[0]!.replace(/^www\d*\./, "");
}

function sameSite(url: string, baseDomain: string): boolean {
  try {
    const host = new URL(url).hostname;
    return registrable(host) === baseDomain || !host;
  } catch {
    return false;
  }
}

function emailRank(email: string): number {
  const local = email.split("@", 1)[0]!.toLowerCase();
  if (FINANCE_PREFIXES.includes(local)) return 0;
  if (FINANCE_PREFIXES.some((p) => local.startsWith(p))) return 1;
  if (GENERIC_PREFIXES.includes(local)) return 2;
  if (GENERIC_PREFIXES.some((p) => local.startsWith(p))) return 3;
  return 4;
}

export interface ContactInfo {
  contact_phone: string | null;          // display format, e.g. "020 7946 0123"
  contact_phones: string[];              // display format, deduped via E.164
  contact_phones_e164: string[];         // canonical "+44…" for the same list, same order
  contact_phone_types: string[];         // "fixed_line" | "mobile" | "toll_free" | "" — same order
  contact_email: string | null;
  contact_emails_found: string[];
  contact_page: string | null;
  about_page: string | null;
  team_page: string | null;
}

export const CONTACT_FIELDS = [
  "contact_phone", "contact_phones", "contact_phones_e164", "contact_phone_types",
  "contact_email", "contact_emails_found",
  "contact_page", "about_page", "team_page",
] as const;

const EMPTY_CONTACT: ContactInfo = {
  contact_phone: null, contact_phones: [], contact_phones_e164: [], contact_phone_types: [],
  contact_email: null, contact_emails_found: [],
  contact_page: null, about_page: null, team_page: null,
};

/** Extract phones / on-page emails / contact-about-team page links from page HTML. */
export function extractContactInfo(html: string, baseUrl: string, companyDomain?: string): ContactInfo {
  const out: ContactInfo = { ...EMPTY_CONTACT };
  if (!html) return out;

  let baseDomain = "";
  try {
    baseDomain = companyDomain ? registrable(companyDomain) : registrable(new URL(baseUrl).hostname);
  } catch { baseDomain = ""; }

  // Reverse Cloudflare obfuscation up-front so cheerio gets clean mailto: text.
  const deobfHtml = deobfuscateCloudflareEmails(html);

  let $: cheerio.CheerioAPI;
  try { $ = cheerio.load(deobfHtml); } catch { return out; }

  type PhoneHit = { display: string; e164: string; type: string };
  const phoneHits: PhoneHit[] = [];
  const emails: string[] = [];

  $('a[href^="tel:"]').each((_i, el) => {
    const p = parsePhone(($(el).attr("href") || "").slice(4));
    if (p) phoneHits.push(p);
  });
  $('a[href^="mailto:"]').each((_i, el) => {
    const e = ($(el).attr("href") || "").slice(7).split("?")[0]!.trim().toLowerCase();
    if (e.includes("@")) emails.push(e);
  });

  $("a[href]").each((_i, el) => {
    const href = ($(el).attr("href") || "").trim();
    if (!href || /^(mailto:|tel:|javascript:|#)/.test(href)) return;
    let absu: string;
    try { absu = new URL(href, baseUrl).toString(); } catch { return; }
    if (baseDomain && !sameSite(absu, baseDomain)) return;
    let path = "";
    try { path = new URL(absu).pathname; } catch { path = ""; }
    const hay = `${path} ${$(el).text().replace(/\s+/g, " ").trim()}`.toLowerCase();
    for (const field of ["contact_page", "about_page", "team_page"] as const) {
      if (out[field]) continue;
      if (PAGE_KEYWORDS[field].some((k) => new RegExp(`(^|[/\\s\\-_])${k.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}([/\\s\\-_]|$)`).test(hay))) {
        out[field] = absu;
      }
    }
  });

  // free-text scan (phones/emails not wrapped in links). Run the [at]/[dot]
  // deobfuscator over plain text so obfuscated emails get caught too.
  const rawText = ($("body").length ? $("body").text() : $.root().text()).replace(/\s+/g, " ");
  const text = deobfuscateAtDot(rawText);
  for (const m of text.match(PHONE_RE) ?? []) { const p = parsePhone(m); if (p) phoneHits.push(p); }
  for (const m of text.match(EMAIL_RE) ?? []) emails.push(m.trim().toLowerCase());

  // dedupe + filter emails
  const seenE = new Set<string>();
  const cleanE: string[] = [];
  for (const e of emails) {
    if (seenE.has(e)) continue;
    const at = e.indexOf("@");
    if (at < 1 || at === e.length - 1) continue;            // no/empty local or host
    const host = e.slice(at + 1);
    if (JUNK_EMAIL_HOSTS.some((j) => host.endsWith(j))) continue;
    if (baseDomain && !(host === baseDomain || host.endsWith(`.${baseDomain}`))) continue;
    seenE.add(e);
    cleanE.push(e);
  }
  cleanE.sort((a, b) => emailRank(a) - emailRank(b));

  // dedupe phones by E.164, keep first-seen order
  const seenP = new Set<string>();
  const display: string[] = [], e164: string[] = [], types: string[] = [];
  for (const p of phoneHits) {
    if (seenP.has(p.e164)) continue;
    seenP.add(p.e164);
    display.push(p.display); e164.push(p.e164); types.push(p.type);
  }

  out.contact_phones = display.slice(0, 3);
  out.contact_phones_e164 = e164.slice(0, 3);
  out.contact_phone_types = types.slice(0, 3);
  out.contact_phone = display[0] ?? null;
  out.contact_emails_found = cleanE.slice(0, 4);
  out.contact_email = cleanE[0] ?? null;
  return out;
}

// ── route suggestion ─────────────────────────────────────────────────────────
interface RouteLead {
  contact_phone?: string | null;
  contact_email?: string | null;
  contact_emails_found?: string[];
  contact_page?: string | null;
  director_name?: string | null;
  director_role?: string | null;
  generic_emails?: Array<{ email?: string }> | null;
  guessed_emails?: Array<{ email?: string }> | null;
  website?: string | null;
}

export function hasContactRoute(lead: RouteLead): boolean {
  return !!(lead.contact_phone || lead.contact_email || (lead.contact_emails_found?.length)
    || lead.contact_page || lead.director_name || (lead.guessed_emails?.length) || lead.website);
}

export function bestContactRoute(lead: RouteLead): string {
  const phone = lead.contact_phone ?? null;
  const emailsFound = lead.contact_emails_found?.length ? lead.contact_emails_found : (lead.contact_email ? [lead.contact_email] : []);
  const contactPage = lead.contact_page ?? null;
  const director = lead.director_name ?? null;
  const role = lead.director_role ?? null;
  const generic = (lead.generic_emails ?? lead.guessed_emails ?? []).map((e) => e?.email).filter((x): x is string => !!x);
  const website = lead.website ?? null;

  const fin = emailsFound.find((e) => e.includes("@") && FINANCE_PREFIXES.includes(e.split("@", 1)[0]!.toLowerCase()));
  if (fin) return `Email ${fin} directly (accounts/finance) — say you'd like a quick word with whoever handles supplier payments / currency.`;
  if (phone) {
    const who = !director ? "the Finance Director / MD" : `${director}${role ? ` (${role})` : ""}`;
    return `Call the switchboard on ${phone} — ask for ${who}.`;
  }
  const realGeneric = emailsFound.find((e) => e.includes("@") && GENERIC_PREFIXES.includes(e.split("@", 1)[0]!.toLowerCase()));
  if (realGeneric) return `Email ${realGeneric} — ask to be put through to whoever handles supplier payments / FX.`;
  if (contactPage) {
    const who = !director ? "the Finance Director / MD" : `${director}${role ? ` (${role})` : ""}`;
    return `Use the contact page (${contactPage}) — ask for ${who}.`;
  }
  if (director) return `Look up ${director}${role ? ` (${role})` : ""} on LinkedIn (or use the Companies-House director route), then call the switchboard.`;
  if (generic.length) return `Find the switchboard number from the website; failing that, try ${generic[0]}.`;
  if (website) return `Pull the phone number / contact page from ${website}.`;
  return "No contact route yet — needs manual research (try Companies House for a director, then the website).";
}

const ARRAY_FIELDS: ReadonlySet<typeof CONTACT_FIELDS[number]> = new Set([
  "contact_phones", "contact_phones_e164", "contact_phone_types", "contact_emails_found",
]);

/** Set the CONTACT_FIELDS to their defaults on a lead-like object (idempotent). */
export function clearContactFields(lead: Record<string, unknown>): void {
  for (const f of CONTACT_FIELDS) lead[f] = ARRAY_FIELDS.has(f) ? [] : null;
}

/** Ensure CONTACT_FIELDS exist on a lead-like object (default null / []), without overwriting set values. */
export function ensureContactFields(lead: Record<string, unknown>): void {
  for (const f of CONTACT_FIELDS) {
    if (lead[f] === undefined) lead[f] = ARRAY_FIELDS.has(f) ? [] : null;
  }
}

File: scripts/contacts.py
"""
contacts.py — pull contact routes out of a company's own website HTML.

No API calls. Given the raw HTML already fetched during website validation,
extract: phone numbers, on-page emails, and links to the contact / about / team
pages. best_contact_route(lead) then turns the lead's combined fields (these +
the Companies-House director + the guessed generic emails) into a one-line
"how to get in front of a decision-maker" suggestion.

Brain 3 responsibility per CLAUDE.md ("Extracts phone, email, contact page";
"Gate 5 — Contact route"). Parses a *fresh* soup from the raw HTML rather than
the cleaned text soup used for signal extraction, so header/footer (where the
phone + contact link usually live) aren't stripped.
"""

import re
from urllib.parse import urljoin, urlparse

from bs4 import BeautifulSoup

# Role-prefix priorities (lower index = better for an FX conversation)
_FINANCE_PREFIXES = ("accounts", "finance", "ap", "ar", "payments", "treasury", "purchasing", "procurement")
_GENERIC_PREFIXES = ("info", "enquiries", "enquiry", "hello", "contact", "office", "admin", "sales", "mail", "team")
_JUNK_EMAIL_HOSTS = ("sentry.io", "wixpress.com", "example.com", "domain.com", "email.com",
                     "yourdomain.com", "sentry-next.wixpress.com")

_PAGE_KEYWORDS = {
    "contact_page": ("contact", "contact-us", "contactus", "get-in-touch"),
    "about_page":   ("about", "about-us", "aboutus", "who-we-are", "our-story", "our-company", "the-company"),
    "team_page":    ("team", "our-team", "meet-the-team", "people", "our-people", "management",
                     "leadership", "directors", "staff", "key-people"),
}

# UK-ish phone numbers in free text: optional +44 / leading 0, then 9-13 more digits
# allowing spaces / brackets / hyphens. Conservative to avoid grabbing dates / IDs.
_PHONE_RE = re.compile(
    r"(?<![\d/])(?:\+44\s?\(?0?\)?\s?|\(?0)\s?\d{1,5}[)\s.-]?\s?\d{2,4}[\s.-]?\d{3,4}(?![\d/])"
)
_EMAIL_RE = re.compile(r"[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}")


def _norm_phone(raw):
    digits = re.sub(r"[^\d+]", "", raw or "")
    digits = re.sub(r"^\+440", "+44", digits)
    digits = re.sub(r"^00", "+", digits)
    n = re.sub(r"\D", "", digits)
    if not (9 <= len(n) <= 13):
        return None
    return re.sub(r"\s+", " ", (raw or "").strip())


def _registrable(netloc):
    netloc = (netloc or "").lower().split(":")[0]
    return re.sub(r"^www\d*\.", "", netloc)


def _same_site(url, base_domain):
    try:
        host = urlparse(url).netloc
        return _registrable(host) == base_domain or not host
    except Exception:
        return False


def _email_rank(email):
    local = email.split("@", 1)[0].lower()
    if local in _FINANCE_PREFIXES:                          return 0
    if any(local.startswith(p) for p in _FINANCE_PREFIXES): return 1
    if local in _GENERIC_PREFIXES:                          return 2
    if any(local.startswith(p) for p in _GENERIC_PREFIXES): return 3
    return 4


def extract_contact_info(html, base_url, company_domain=None):
    """
    html          : raw HTML string of the homepage
    base_url      : the (final) URL the HTML came from — used to resolve relative links
    company_domain: bare domain (e.g. "acmewine.co.uk"); if omitted, derived from base_url.
                    Only on-page emails / page links on this domain are kept.
    Returns a dict of contact fields (all keys always present).
    """
    out = {
        "contact_phone": None, "contact_phones": [],
        "contact_email": None, "contact_emails_found": [],
        "contact_page": None, "about_page": None, "team_page": None,
    }
    if not html:
        return out

    try:
        base_domain = _registrable(company_domain) if company_domain else _registrable(urlparse(base_url).netloc)
    except Exception:
        base_domain = ""

    try:
        soup = BeautifulSoup(html, "html.parser")
    except Exception:
        return out

    phones, emails = [], []

    for a in soup.select('a[href^="tel:"]'):
        p = _norm_phone(a.get("href", "")[4:])
        if p:
            phones.append(p)
    for a in soup.select('a[href^="mailto:"]'):
        e = a.get("href", "")[7:].split("?")[0].strip().lower()
        if "@" in e:
            emails.append(e)

    for a in soup.find_all("a", href=True):
        href = a["href"].strip()
        if href.startswith(("mailto:", "tel:", "javascript:", "#")):
            continue
        absu = urljoin(base_url, href)
        if base_domain and not _same_site(absu, base_domain):
            continue
        hay = (urlparse(absu).path + " " + (a.get_text(" ", strip=True) or "")).lower()
        for field, kws in _PAGE_KEYWORDS.items():
            if out[field]:
                continue
            if any(re.search(rf"(^|[/\s\-_]){re.escape(k)}([/\s\-_]|$)", hay) for k in kws):
                out[field] = absu

    # free-text scan (catches phones/emails not wrapped in links)
    text = re.sub(r"\s+", " ", soup.get_text(" ", strip=True))
    for m in _PHONE_RE.findall(text):
        p = _norm_phone(m)
        if p:
            phones.append(p)
    for m in _EMAIL_RE.findall(text):
        emails.append(m.strip().lower())

    seen_e, clean_e = set(), []
    for e in emails:
        if e in seen_e or "@" not in e:
            continue
        host = e.split("@", 1)[1]
        if any(host.endswith(j) for j in _JUNK_EMAIL_HOSTS):
            continue
        if base_domain and not (host == base_domain or host.endswith("." + base_domain)):
            continue
        seen_e.add(e)
        clean_e.append(e)
    clean_e.sort(key=_email_rank)

    seen_p, clean_p = set(), []
    for p in phones:
        key = re.sub(r"\D", "", p)
        if not key or key in seen_p:
            continue
        seen_p.add(key)
        clean_p.append(p)

    out["contact_phones"]       = clean_p[:3]
    out["contact_phone"]        = clean_p[0] if clean_p else None
    out["contact_emails_found"] = clean_e[:4]
    out["contact_email"]        = clean_e[0] if clean_e else None
    return out


# ── ROUTE SUGGESTION ─────────────────────────────────────────────────────────────

CONTACT_FIELDS = ("contact_phone", "contact_phones", "contact_email",
                  "contact_emails_found", "contact_page", "about_page", "team_page")


def has_contact_route(lead):
    return bool(
        lead.get("contact_phone") or lead.get("contact_email") or lead.get("contact_emails_found")
        or lead.get("contact_page") or lead.get("director_name") or lead.get("guessed_emails")
        or lead.get("website")
    )


def best_contact_route(lead):
    phone        = lead.get("contact_phone")
    emails_found = lead.get("contact_emails_found") or ([lead["contact_email"]] if lead.get("contact_email") else [])
    contact_page = lead.get("contact_page")
    director     = lead.get("director_name")
    role         = lead.get("director_role")
    generic      = [e.get("email") for e in (lead.get("generic_emails") or lead.get("guessed_emails") or []) if isinstance(e, dict) and e.get("email")]
    website      = lead.get("website")

    fin = next((e for e in emails_found if "@" in e and e.split("@", 1)[0].lower() in _FINANCE_PREFIXES), None)
    if fin:
        return (f"Email {fin} directly (accounts/finance) — say you'd like a quick word with whoever "
                f"handles supplier payments / currency.")
    if phone:
        who = "the Finance Director / MD" if not director else (f"{director}" + (f" ({role})" if role else ""))
        return f"Call the switchboard on {phone} — ask for {who}."
    real_generic = next((e for e in emails_found if "@" in e and e.split("@", 1)[0].lower() in _GENERIC_PREFIXES), None)
    if real_generic:
        return f"Email {real_generic} — ask to be put through to whoever handles supplier payments / FX."
    if contact_page:
        who = "the Finance Director / MD" if not director else (f"{director}" + (f" ({role})" if role else ""))
        return f"Use the contact page ({contact_page}) — ask for {who}."
    if director:
        return (f"Look up {director}" + (f" ({role})" if role else "") +
                " on LinkedIn (or use the Companies-House director route), then call the switchboard.")
    if generic:
        return f"Find the switchboard number from the website; failing that, try {generic[0]}."
    if website:
        return f"Pull the phone number / contact page from {website}."
    return "No contact route yet — needs manual research (try Companies House for a director, then the website)."
